src/Orleans.Core.Abstractions/Streams/Core/IStreamSubscriptionHandleFactory.cs

```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Orleans.Runtime;
+
+namespace Orleans.Streams.Core
+{
+    public interface IStreamSubscriptionHandleFactory
+    {
+        StreamSubscriptionHandle<T> Create<T>(GuidId subscriptionId, IStreamIdentity streamId, string streamProviderName);
```

Comment:
Only thing we need from the application is the type (T).
And the application needs know only the stream identity and provider to make the decision if it should subscribe or not.

So Is suggest:

    public interface IStreamSubscriptionHandleFactory 
    {
        IStreamIdentity StreamId { get; }
        string ProviderName { get; }
        StreamSubscriptionHandle<T> Create<T>();
    }

This would make
 
    public interface IStreamSubscriptionObserver
    { 
        Task OnSubscribed(IStreamSubscriptionHandleFactory handleFactory);
    }

The advantage of this is that it provides less room for error.  In the previous pattern there is nothing prohibiting the user from passing in incorrect streamed, subscriptionid or provider name.